python/cudf/cudf/utils/utils.py

```diff
@@ -37,6 +39,65 @@
 }
 
 
+# The test root is set by pytest to support situations where tests are run from
+# a source tree on a built version of cudf.
+NO_EXTERNAL_ONLY_APIS = os.getenv("NO_EXTERNAL_ONLY_APIS")
+_CUDF_TEST_ROOT = os.getenv("_CUDF_TEST_ROOT")
+
+if NO_EXTERNAL_ONLY_APIS in ("True", "1", "TRUE"):
+    _cudf_root = cudf.__file__.replace("/__init__.py", "")
+    # If the environment variable for the test root is not set, we default to
+    # using the path relative to the cudf root directory.
+    _tests_root = _CUDF_TEST_ROOT or os.path.join(_cudf_root, "tests")
+
+    def _external_only_api(func, alternative=""):
+        """Decorator to indicate that a function should not be used internally.
+
+        cudf contains many APIs that exist for pandas compatibility but are
+        intrinsically inefficient. For some of these cudf has internal
+        equivalents that are much faster. Usage of the slow public APIs inside
+        our implementation can lead to unnecessary performance bottlenecks.
+        Applying this decorator to such functions and then setting the
+        environment variable NO_EXTERNAL_ONLY_APIS will cause such functions to
+        raise exceptions if they are called from anywhere inside cudf, making
+        it easy to identify and excise such usage.
+        """
+
+        # If the first arg is a string then an alternative function to use in
+        # place of this API was provided, so we pass that to a subsequent call.
+        # It would be cleaner to implement this pattern by using a class
+        # decorator with a factory method, but there is no way to generically
+        # wrap docstrings on a class (we would need the docstring to be on the
+        # class itself, not instances, because that's what `help` looks at) and
+        # there is also no way to make mypy happy with that approach.
+        if isinstance(func, str):
+            return lambda actual_func: _external_only_api(actual_func, func)
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            # Check the immediately preceding frame to see if it's in cudf.
+            frame, lineno = next(traceback.walk_stack(None))
+            fn = frame.f_code.co_filename
+            if _cudf_root in fn and _tests_root not in fn:
+                raise RuntimeError(
+                    f"External-only API called in {fn} at line {lineno}. "
+                    f"{alternative}"
+                )
+            return func(*args, **kwargs)
+
+        return wrapper
+
+
+else:
```

Comment:
Should the environment variable check be _inside_ the function definition, and trigger early exit? e.g.
```python
def _external_only_api(func, alternative=""):
    """Docstring."""
    if isinstance(func, str):
        return lambda actual_func: _external_only_api(actual_func, func)
    if not NO_EXTERNAL_ONLY_APIS:  # Ignore the double negative here, it's fine. ;)
        return func
    # insert fancy wrapper / etc.
    return wrapper
```

Reply:
Sure, we can do that instead. Prior to adding the `alternative` the default implementation was really a no-op, so conditionally defining the function led to zero overhead on import time. Now I suppose that we may as well inline this check, the extra os calls are basically negligible in the total cudf import time.